### PR TITLE
Lower

### PR DESCRIPTION
**Pull request by joostvanzwieten**

### CODE REVIEW

File: nutils/function.py [hunk 1]
```diff
@@ -37,7 +37,26 @@
...
+        The dimension of the :class:`~nutils.sample.Sample` on which the
+        resulting :class:`nutils.evaluable.Array` will be evaluated.
+    opposite : :class:`bool`
+        Indicates which transform chain to use when evaluating the resulting
+        :class:`~nutils.evaluable.Array`. This has no effect when there is only
+        one transform chain.
+    with_points : :class:`bool`
+        If true the resulting :class:`~nutils.evaluable.Array` will have a leading points axis.
+    npoints : :class:`int` or :class:`nutils.evaluable.Array`
+        The length of the points axis.
```
Comment (gertjanvanzwieten) [inline nutils/function.py:55]: can't this be combined with with_points?

File: nutils/function.py [hunk 2]
```diff
@@ -626,7 +663,7 @@ def __init__(self, ndims: int) -> None:
   def __getnewargs__(self):
     return self.shape[0],
 
-  def prepare_eval(self, **kwargs: Any) -> evaluable.Array:
+  def prepare_eval(self, *, with_points: bool = True, **kwargs: Any) -> evaluable.Array:
```
Comment (gertjanvanzwieten) [inline nutils/function.py:666]: what is the point of adding with_points here?

**Review by gertjanvanzwieten — COMMENTED** [review]
Some minor things for now; I still need to study the new function.py in detail. Looks great though!

**Review by gertjanvanzwieten — APPROVED** [review]

File: nutils/function.py [hunk 3]
```diff
@@ -18,2906 +18,2514 @@
...
-      try:
-        f, args = obj.__reduce__()
-      except: # obj cannot be reduced into a constructor and its arguments
-        rstack.append(obj)
-      else:
-        fstack.append(f)
-        fstack.append(len(args))
-        fstack.append(recreate)
-        fstack.extend(args)
+      raise ValueError
```
Comment (gertjanvanzwieten) [inline nutils/function.py:118]: Let's add a message here such as ValueError('cannot convert {} to Array'.format(type(__value).__name). Also perhaps we can switch to single underscores for __value so it becomes printable in pdb.

**Review by gertjanvanzwieten — CHANGES_REQUESTED** [review]

File: nutils/function.py (hunk 3 above)
Comment (joostvanzwieten) [reply nutils/function.py]: Done. Mypy treats parameters with double underscores as positional only (see https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html#functions , function `quux`).

Comment (gertjanvanzwieten) [reply nutils/function.py]: Ah yes, that was it.

**Review by gertjanvanzwieten — APPROVED** [review]

**Pull request merged**